server/src/auth/mod.rs

```diff
@@ -0,0 +1,142 @@
+use error::{LoginError, RegisterError, VerifyError};
+
+use crate::models::{DatedToken, Token, User};
+use crate::repository::{MetaRepo, Users};
+
+pub mod error;
+pub mod middleware;
+
+/// Gets a user based on their token. Verifies that they are logged in and that their token is not
+/// yet expired.
+///
+/// ```
+/// # use big_cheese_server::test_util::test_db;
+/// use big_cheese_server::auth::{register, login, verify};
+/// # test_db(|client| async move {
+///     # client.meta_repo.set_default_meta_if_not_exists().await.unwrap();
+///
+///     let users = &client.user_repo;
+///     let meta = &client.meta_repo;
+///
+///     register(users, "test", "yeet").await.unwrap();
+///     let token = login(users, "test", "yeet").await.unwrap();
+///
+///     let user = verify(users, meta, token).await.unwrap();
+///     assert_eq!(user.username, "test");
+/// # });
+/// ```
+///
+/// # Expiration:
+/// ```
+/// # use big_cheese_server::test_util::test_db;
+/// use big_cheese_server::auth::{register, login, verify};
+/// use big_cheese_server::models::Config;
+/// use std::time::Duration;
+/// use std::thread;
+/// use big_cheese_server::auth::error::VerifyError;
+/// # test_db(|client| async move {
+///     # client.meta_repo.set_default_meta_if_not_exists().await.unwrap();
+///
+///     let users = &client.user_repo;
+///     let meta = &client.meta_repo;
+///
+///     meta.set_config(Config {expiration_time: Duration::from_millis(1)}).await.unwrap();
+///
+///     register(users, "test", "yeet").await.unwrap();
+///     let token = login(users, "test", "yeet").await.unwrap();
+///
+///     // let the token expire
+///     thread::sleep(Duration::from_millis(100));
+///
+///     assert!(matches!(verify(users, meta, token).await, Err(VerifyError::Expired)));
+/// # });
+/// ```
+pub async fn verify(users: &Users, meta: &MetaRepo, token: Token) -> Result<User, VerifyError> {
+    let user = users
+        .get_user_for_token(token)
+        .await?
+        .ok_or(VerifyError::Invalid)?;
+
+    let config = meta.get_config().await?;
+
+    let dt = user
+        .tokens
+        .iter()
+        .find(|t| t.token == token)
+        .ok_or(VerifyError::Invalid)?;
+
+    // TODO: test
```

Comment:
Should this be in this PR or another. If in another making an issue would be nice

Reply:
More testing is an issue in general. I have started doing some testing in this PR though. I'll make an issue for more.